### PR TITLE
add substitution for functional linear

### PR DESCRIPTION
**Pull request by itai-berman**
## Pull Request Description:
Add substitution support from functional linear layer to Linear in Pytorch.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core import common
+from model_compression_toolkit.core.common import BaseNode, Graph
+from model_compression_toolkit.core.common.graph.functional_node import FunctionalNode
+from model_compression_toolkit.core.pytorch.constants import *
+from model_compression_toolkit.logger import Logger
+
+
+class FunctionalLinear(common.BaseSubstitution):
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py:26]: just import BaseSubstitution from common

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    """
+    Replace functional layer_norm with LayerNorm.
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py:28]: layer_norm?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        Returns:
+            Graph after applying the substitution.
+        """
+
+        # Create new node of layer convolution
+        if 1 not in func_node.weights:
+            Logger.critical(f'Weight input missing for node {func_node.name}.')  # pragma: no cover
+
+        weight = func_node.weights[1]
+        bias = func_node.weights.get(2, None)
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py:56]: make sure the bias positional weight is saved as 2 even when the linear is called with bias as kwarg

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            BIAS: bias is not None,
+        }
+
+        weights = {KERNEL: weight} if bias is None else {KERNEL: weight, BIAS: bias}
+
+        new_node = BaseNode(
+            name=func_node.name,
+            framework_attr=framework_attr,
+            input_shape=func_node.input_shape[0],
+            output_shape=tuple(func_node.output_shape[0]),
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py:70]: why not just `output_shape=func_node.output_shape`?

File: tests/pytorch_tests/model_tests/feature_models/linear_function_test.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+class LinearFNet(torch.nn.Module):
+    def __init__(self):
+        super(LinearFNet, self).__init__()
+        self.fc1 = torch.nn.Linear(in_features=1000, out_features=100, bias=False)
+        self.fc2 = torch.nn.Linear(in_features=100, out_features=10, bias=True)
+
+    def forward(self, x):
+        x = F.linear(x, self.fc1.weight, self.fc1.bias)
+        y = F.linear(x, self.fc2.weight, self.fc2.bias)
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/linear_function_test.py:33]: change to:
`y = F.linear(x, bias=self.fc2.bias, weight=self.fc2.weight)`
to make sure everything still works

**Review by elad-c — COMMENTED** [review]
what about `torch.nn.functional.conv2d` and it's friends?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py (hunk 1 above)
Comment (itai-berman) [reply model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/functional_linear.py:26]: I just copied the code from all other substitutions. Should I change it?

**Comment by itai-berman** [comment]
It seems that you already handled both Conv2d and ConvTranspose2d `FunctionalConvSubstitution` and also `FunctionalBatchNorm`, `FunctionalLayerNorm` were handled

File: tests/pytorch_tests/model_tests/feature_models/linear_function_test.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+class LinearFNet(torch.nn.Module):
+    def __init__(self):
+        super(LinearFNet, self).__init__()
+        self.fc1 = torch.nn.Linear(in_features=1000, out_features=100, bias=False)
+        self.fc2 = torch.nn.Linear(in_features=100, out_features=10, bias=True)
+
+    def forward(self, x):
+        x = F.linear(x, self.fc1.weight, self.fc1.bias)
+        y = F.linear(x, bias=self.fc2.bias, weight=self.fc2.weight)
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/linear_function_test.py:33]: add a conv with bias=None

**Review by elad-c — APPROVED** [review]
